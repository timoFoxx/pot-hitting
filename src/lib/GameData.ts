export enum Dimensions {
    height = 600,
    width = 500,
}

export enum PotCoords {
    potX = Math.random() * Dimensions.height,
    potY = Math.random() * Dimensions.width,
}
