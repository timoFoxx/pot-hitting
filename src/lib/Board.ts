import { Point } from "./Game";


export type potX = number;

export type potY = number;

export class Board {
    // Upper left corner point is denoted by x and y
    potWidth: number = 20
    potHeight: number = 20

    constructor(public x: potX, public y: potY) {}

    isOnPot = (p: Point): Boolean => {
        return Math.abs(p.x - this.x) < this.potWidth && Math.abs(p.y - this.y) < this.potHeight
    }

}
