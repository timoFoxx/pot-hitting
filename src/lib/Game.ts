import { Board } from './Board'

export class Point {
    constructor(public x: number, public y: number) {}

    //calculates distance from mouse pointer to pot coords
    distanceFrom = (p: Point): number  => Math.sqrt(Math.pow(this.x - p.x, 2) + Math.pow(this.y - p.y, 2))
}

export interface GameEvent {
    score: number;
}

export class Win implements GameEvent {
    constructor(public score: number) {}
}

export class Cold implements GameEvent {
    constructor(public score: number) {}
}

export class Warm implements GameEvent {
    constructor(public score: number) {}
}

export class FirstClick implements GameEvent {
    constructor(public score: number) {}
}

//side effects
export class GameState {
    constructor(public distance: number | null, public score: number) {}

    setDistance = (p: number) => {
        this.distance = p
    }

    incrementScore = () => {
        this.score += 1;
    }
}

export type GameLoop = (e: Point) => GameEvent

export const Game = (board: Board, state: GameState = new GameState(null, 0)): GameLoop => (userEvent: Point): GameEvent => {
    const newDistance = userEvent.distanceFrom(new Point(board.x, board.y))
    // side effects here -> redux?
    if (board.isOnPot(userEvent)) {
        state.incrementScore()
        return new Win(state.score)
    } else if (state.distance == null) {
        state.setDistance(newDistance)
        return new FirstClick(state.score)
    } else if (state.distance > newDistance) {
        state.setDistance(newDistance)
        state.incrementScore()
        return new Warm(state.score)
    } else {
        state.setDistance(newDistance)
        state.incrementScore()
        return new Cold(state.score)
    }
}
