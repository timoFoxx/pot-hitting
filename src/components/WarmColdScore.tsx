import React from 'react'

import {Cold, GameEvent, Warm, Win} from '../lib/Game'
import {StyledMessageContainer, StyledMessageContainerCold, StyledMessageContainerWarm} from "../styles/Layout";

interface ScoreProps {
    event: GameEvent | null
}

export const WarmColdScore = ({ event} : ScoreProps) => {
    if (event instanceof Win) {
        return (<StyledMessageContainer>Congratulations, You scored amazingly good, like {event.score}!</StyledMessageContainer>)
    } else if (event instanceof Cold) {
        return (<StyledMessageContainerCold>Cold</StyledMessageContainerCold>)
    } else if (event instanceof Warm) {
        return (<StyledMessageContainerWarm>Warm</StyledMessageContainerWarm>)
    } else {
        return null
    }
}
