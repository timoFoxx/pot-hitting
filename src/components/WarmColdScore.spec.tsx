import React from 'react'
import renderer from 'react-test-renderer'
import { Win } from '../lib/Game'

import { WarmColdScore} from "./WarmColdScore";

describe('WarmColdScore', () => {
    test('should render according to the snapshot when won', () => {
        expect(renderer.create(<WarmColdScore event={new Win(0)} />).toJSON()).toMatchSnapshot()
    })
})
