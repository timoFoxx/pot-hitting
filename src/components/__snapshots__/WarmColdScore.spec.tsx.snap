// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WarmColdScore should render according to the snapshot when won 1`] = `
<div
  className="sc-hKFyIo sc-gKAblj cUVHgQ frNQAl"
>
  Congratulations, You scored amazingly good, like 
  0
  !
</div>
`;
